tracer/src/Datadog.Trace/Logging/TracerFlare/DebugLogReader.cs

```diff
@@ -0,0 +1,94 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.IO;
+using System.IO.Compression;
+using System.Threading.Tasks;
+
+namespace Datadog.Trace.Logging.TracerFlare;
+
+internal static class DebugLogReader
+{
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(DebugLogReader));
+
+    public static bool TryToCreateSentinelFile(string logDirectory, string flareRequestId)
+    {
+        // Assuming the log directory is accessible and writeable
+        // If it isn't this will return false anyway, but we do an extra check, just to be sure
+        try
+        {
+            // Using the dotnet-tracer-sentinel-*.log pattern means it gets automatically cleaned up by the log cleaning task
+            var filePath = Path.Combine(logDirectory, $"dotnet-tracer-sentinel-{flareRequestId}.log");
+            using var file = File.Open(filePath, FileMode.CreateNew, FileAccess.Write);
+
+            // if we got this far, the file was created
+            return true;
+        }
+        catch (IOException)
+        {
+            // expected, the file already exists
+            Log.Information("Sentinel file for config {FlareRequestID} found in {LogDirectory}", flareRequestId, logDirectory);
+            return false;
+        }
+        catch (Exception ex)
+        {
+            // check if the log directory exists, to see if this is an "expected" consequence
+            if (Directory.Exists(logDirectory))
+            {
+                // strange error
+                Log.Warning(ex, "Error creating sentinel file for config {FlareRequestID} in {LogDirectory}", flareRequestId, logDirectory);
+            }
+            else
+            {
+                Log.Warning(ex, "Error creating sentinel file for config {FlareRequestID} in {LogDirectory}: Directory does not exist", flareRequestId, logDirectory);
+            }
+
+            return false;
+        }
+    }
+
+    public static async Task WriteDebugLogArchiveToStream(Stream writeTo, string logDirectory)
+    {
+        try
+        {
+            using var archive = new ZipArchive(writeTo, ZipArchiveMode.Create, true);
+
+            // Only sending .log files to avoid the risk of sending files we don't want.
+            // Also not recurrsing, in-case they have a weird log setup
+            // Grabbing all the filenames once, as we could be creating more in the background, and don't
+            // want to run into any modified enumerable issues etc
+            var filesToUpload = Directory.GetFiles(logDirectory, "*.log", SearchOption.TopDirectoryOnly);
+            if (filesToUpload.Length == 0)
+            {
+                // no files to upload
+                return;
+            }
+
+            foreach (var filePath in filesToUpload)
```

Comment:
I think we have to be smarter here. The zendesk attachment has a 30MB limit. I think we should process the files in reverse chronological order, possibly limit ourselves to x days, and stop when archive gets near a threshold.

Logs are fairly predictable, so I think we can make an educated guess of what the compression ratio is going to be (and then add some margin).

Reply:
What exactly happens if it somehow exceeds the 30MB limit?

Reply:
As I think there is still the potentially rare possibility that the compression ratio estimate for a log file is incorrect since it can't validate it prior to writing to the ZipArchive

Reply:
@kevingosse can confirm, but I think it would basically get rejected by the zendesk header. We've done some testing of real log directories, and we still get 12/13x compression, so we'll _probably_ be fine.

Unfortunately, as we're streaming this directly in the request, it's not like we can "restart" or anything, so I think it's probably the best we can do 🤷‍♂️